File: NC/three-integer-sum.py
class Solution:

    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()

        def binSearch(start: int, end: int, prevSum: int):
            if start > end:
                return -1

            curr = int((start+end)/2)

            sum = prevSum+nums[curr]

            if sum == 0:
                return curr
            elif sum > 0:
                return binSearch(start, curr-1, prevSum)
            else:
                return binSearch(curr+1, end, prevSum)

        ans = set()

        print(nums)

        for i in range(len(nums)):
            for j in range(len(nums)-1, i, -1):
                k = binSearch(i+1, j-1, nums[i]+nums[j])
                print(nums[i], nums[k], nums[j])
                if k != -1:
                    ans.add((nums[i], nums[k], nums[j]))

        return [[i[0], i[1], i[2]] for i in ans]

File: NC/three-integer-sum.cpp
class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        sort(nums.begin(), nums.end());

        vector<vector<int>> ans;

        for (int curr = 0; curr < nums.size(); curr++)
        {
            if (curr > 0 && nums[curr - 1] == nums[curr])
            {
                continue;
            }
            int target = nums[curr];

            int i = curr + 1, j = nums.size() - 1;

            while (i < j)
            {
                int leftNum = nums[i];
                int rightNum = nums[j];
                int ansLength = ans.size();
                if (target + leftNum + rightNum == 0)
                {

                    if (ansLength == 0 ||
                        !(target == ans[ansLength - 1][0] && leftNum == ans[ansLength - 1][1] && rightNum == ans[ansLength - 1][2]))
                    {
                        ans.push_back({target, leftNum, rightNum});
                    }
                    i++;
                    continue;
                }
                else if (target + leftNum + rightNum < 0)
                {
                    i++;
                    continue;
                }
                else
                {
                    j--;
                }
            }
        }

        return ans;
    }
};

// [-1,0,1,2,-1,-4]
// [-4,-1,-1,0,1,2]